README.md
# avtool

[![License: MIT](https://img.shields.io/badge/License-GPL_v3-brightgreen.svg)](https://github.com/clok/avtool/blob/main/LICENSE)
[![Go Report Card](https://goreportcard.com/badge/clok/avtool)](https://goreportcard.com/report/clok/avtool)
[![Coverage Status](https://coveralls.io/repos/github/clok/avtool/badge.svg?branch=main)](https://coveralls.io/github/clok/avtool?branch=main)
[![go.dev reference](https://img.shields.io/badge/go.dev-reference-007d9c?logo=go&logoColor=white)](https://pkg.go.dev/github.com/clok/avtool/v3?tab=overview)

> NOTE: Original code written by [@pbthorste](https://github.com/pbthorste) for [https://github.com/pbthorste/avtool](https://github.com/pbthorste/avtool)
>
> HUGE SHOUT OUT to [@pbthorste](https://github.com/pbthorste)

This module provides a reimplementation of `ansible-vault` encrypt and decrypt functionality in Go.

## CLI Tool

Please see [gwvault](https://github.com/GoodwayGroup/gwvault) for a purpose built `ansible-vault` binary written in go.

It leverages the work done by [@pbthorste](https://github.com/pbthorste) for [https://github.com/pbthorste/avtool](https://github.com/pbthorste/avtool) while further fleshing out the CLI tool to be more in line with the original `ansible-vault` CLI tool.

## v2 to v3

In the `v3.0.0` update there are breaking changes to the interface.

The `Encrypt` and `Decrypt` methods now expect the `EncryptOptions` and `DecryptOptions` parameter types as inputs.

The `EncryptFile` and `DecryptFile` methods now expect the `EncryptFileOptions` and `DecryptFileOptions` parameter types as inputs.

This change was made to allow for easier extension of the library without the need for future major version bumps.

## Why the fork?

As of writing the mainline has been stale for ~4 years.

I have found this code to be highly useful and important for writing other `ansible-vault` related tools. I wanted to modernize the work done previously to support `go.mod` while also updating the interface as an importable module for other code.

## Thanks and Attribution

Original code written by [@pbthorste](https://github.com/pbthorste)


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="v3.0.0"></a>
## [v3.0.0] - 2021-12-21
### Bug Fixes
- **deps:** update golang.org/x/crypto commit hash to 089bfa5 ([#8](https://github.com/clok/avtool/issues/8))

### Chore
- **deps:** update coverallsapp/github-action action to v1.1.3 ([#9](https://github.com/clok/avtool/issues/9))

### Features
- **structured options:** refactor methods to accept structured options and remove all panics ([#13](https://github.com/clok/avtool/issues/13))


<a name="v2.2.0"></a>
## [v2.2.0] - 2021-12-20
### Chore
- update changelog for v2.2.0

### Features
- **vault 1.2:** added support for the ansible vault v1.2 payload format ([#10](https://github.com/clok/avtool/issues/10))


<a name="v2.1.1"></a>
## [v2.1.1] - 2021-05-13
### Chore
- update changelog for v2.1.1


<a name="v2.1.0-2-g1dd422f"></a>
## [v2.1.0-2-g1dd422f] - 2021-05-13
### Bug Fixes
- **deps:** update golang.org/x/crypto commit hash to cd7d49e ([#7](https://github.com/clok/avtool/issues/7))

### Chore
- update changelog for v2.1.0-2-g1dd422f
- **ci:** added Makefile and goreleaser


<a name="v2.1.0"></a>
## [v2.1.0] - 2021-03-15
### Chore
- **ci:** port to using golangci-lint github action
- **deps:** update actions/setup-go action to v2 ([#4](https://github.com/clok/avtool/issues/4))
- **deps:** update actions/checkout action to v2 ([#3](https://github.com/clok/avtool/issues/3))
- **go.mod:** bump to go 1.16

### Features
- **release:** v2.1.0


<a name="v2.0.3"></a>
## [v2.0.3] - 2021-03-03
### Chore
- **README:** update badges

### Features
- **release:** v2.0.3


<a name="v2.0.2"></a>
## [v2.0.2] - 2021-03-03
### Bug Fixes
- **github actions:** point to forked go-proxy-pull-action to support major versions

### Features
- **release:** v2.0.2


<a name="v2.0.1"></a>
## [v2.0.1] - 2021-03-03
### Chore
- **github actions:** update github actions versions
- **github actions:** add go proxy warming

### Features
- **release:** v2.0.1


<a name="v2.0.0"></a>
## [v2.0.0] - 2021-03-02
### Chore
- update README.md
- add CONTRIBUTING.md
- **changelog:** added CHANGELOG.md generated by git-chglog
- **cli:** remove deprecated CLI tool and update README.md
- **lint:** clean up linting issues
- **lint:** clean up code
- **lint:** update for go 1.15 linting rules
- **renovate:** update renovate config
- **tests:** updated tests and added github actions

### Features
- **gomod:** add go.mod support
- **release:** v2.0.0
- **v2:** bump module version to v2


<a name="1.2.0"></a>
## [1.2.0] - 2017-05-25
### Added
- output flag to support vanilla (default) & aligned outputs;
- features to view specific keys/all keys & key names

### Minor
- changes to .gitignore, makefile

### Update
- readme

### Pull Requests
- Merge pull request [#4](https://github.com/clok/avtool/issues/4) from sundarsrd/master


###### Squashed Commits:
```
Added -k flag for selective viewing & additional validations
```



<a name="1.1.1"></a>
## [1.1.1] - 2017-04-21
### Update
- key


<a name="1.1.0"></a>
## [1.1.0] - 2017-04-21
### Add
- travis ci file
- encryption

### Added
- makefile

### Support
- windows line endings

### Use
- version from git


<a name="1.0.0"></a>
## 1.0.0 - 2017-01-24
### Bug Fixes
- usage
- retrievePassword function

### Change
- to GPL v3

### Fm
- t

### Handle
- panic

### Initial
- commit

### Initial
- commit

### Move
- to own function


[Unreleased]: https://github.com/clok/avtool/compare/v3.0.0...HEAD
[v3.0.0]: https://github.com/clok/avtool/compare/v2.2.0...v3.0.0
[v2.2.0]: https://github.com/clok/avtool/compare/v2.1.1...v2.2.0
[v2.1.1]: https://github.com/clok/avtool/compare/v2.1.0-2-g1dd422f...v2.1.1
[v2.1.0-2-g1dd422f]: https://github.com/clok/avtool/compare/v2.1.0...v2.1.0-2-g1dd422f
[v2.1.0]: https://github.com/clok/avtool/compare/v2.0.3...v2.1.0
[v2.0.3]: https://github.com/clok/avtool/compare/v2.0.2...v2.0.3
[v2.0.2]: https://github.com/clok/avtool/compare/v2.0.1...v2.0.2
[v2.0.1]: https://github.com/clok/avtool/compare/v2.0.0...v2.0.1
[v2.0.0]: https://github.com/clok/avtool/compare/1.2.0...v2.0.0
[1.2.0]: https://github.com/clok/avtool/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/clok/avtool/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/clok/avtool/compare/1.0.0...1.1.0


decrypt_test.go
package avtool

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test_Decrypt_V11(t *testing.T) {
	password := []byte("asdf")
	content := []byte(`$ANSIBLE_VAULT;1.1;AES256
39663038636438383965366163636163376531336238346239623934393436393938656439643133
3638363066366433666438623138373866393763373265320a366635386630336562633763323236
61616562393964666464653532636436346535616566613434613361303734373734383930323661
6664306264366235630a643235323438646132656337613434396338396335396439346336613062
3766
`)
	result, err := Decrypt(&DecryptOptions{
		Data:     &content,
		Password: &password,
	})
	assert.Equal(t, "hello", result)
	assert.NoError(t, err)
}

func Test_Decrypt_V12(t *testing.T) {
	password := []byte("asdf")
	content := []byte(`$ANSIBLE_VAULT;1.2;AES256;label
39663038636438383965366163636163376531336238346239623934393436393938656439643133
3638363066366433666438623138373866393763373265320a366635386630336562633763323236
61616562393964666464653532636436346535616566613434613361303734373734383930323661
6664306264366235630a643235323438646132656337613434396338396335396439346336613062
3766
`)
	result, err := Decrypt(&DecryptOptions{
		Data:     &content,
		Password: &password,
	})
	assert.Equal(t, "hello", result)
	assert.NoError(t, err)
}

func Test_DecryptFile(t *testing.T) {
	password := []byte("asdf")
	filename := "./testdata/test1/secrets.yaml"
	_, err := DecryptFile(&DecryptFileOptions{
		Filename: filename,
		Password: &password,
	})
	assert.NoError(t, err)
}

func Test_splitHeader(t *testing.T) {
	content := []byte(`$ANSIBLE_VAULT;1.2;AES256;label
39663038636438383965366163636163376531336238346239623934393436393938656439643133
3638363066366433666438623138373866393763373265320a366635386630336562633763323236
61616562393964666464653532636436346535616566613434613361303734373734383930323661
6664306264366235630a643235323438646132656337613434396338396335396439346336613062
3766
`)
	expected := "396630386364383839653661636361633765313362383462396239343934363939386564396431333638363066366433666438623138373866393763373265320a366635386630336562633763323236616165623939646664646535326364363465356165666134346133613037343737343839303236616664306264366235630a6432353234386461326563376134343963383963353964393463366130623766"
	body := splitHeader(content)
	assert.Equal(t, expected, body)
}

func Test_splitHeader_unsupported_cipher(t *testing.T) {
	content := []byte(`$ANSIBLE_VAULT;1.2;AES128;label
NOOP
`)
	assert.PanicsWithError(t, "unsupported cipher: AES128", func() {
		_ = splitHeader(content)
	})
}

func Test_decodeData_error_bad_hex(t *testing.T) {
	body := "z396630386364383839653661636361633765313362383462396239343934363939386564396431333638363066366433666438623138373866393763373265320d0a3666353866303365626337633232366161656239396466646465353263643634653561656661343461337a0d0a6432353234386461326563376134343963383963353964393463366130623766"
	_, _, _, err := decodeData(body)
	assert.Error(t, err)
}

func Test_decodeData_error_salt(t *testing.T) {
	body := "61206261642073616c740a3132333435360a313233343536"
	_, _, _, err := decodeData(body)
	assert.Error(t, err)
}

func Test_decodeData_error_cryptedHmac(t *testing.T) {
	body := "3132333435360a61206261642063727970740a313233343536"
	_, _, _, err := decodeData(body)
	assert.Error(t, err)
}

func Test_decodeData_error_cipherText(t *testing.T) {
	body := "3132333435360a3132333435360a612062616420636970686572"
	_, _, _, err := decodeData(body)
	assert.Error(t, err)
}


encrypt_test.go
package avtool

import (
	"errors"
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test_Encrypt_V11(t *testing.T) {
	password := []byte("asdf")
	body := []byte("secret")
	var encrypted string
	var err error
	encrypted, err = Encrypt(&EncryptOptions{
		Body:     &body,
		Password: &password,
	})
	assert.NoError(t, err)

	var result string
	data := []byte(encrypted)
	result, err = Decrypt(&DecryptOptions{
		Data:     &data,
		Password: &password,
	})
	assert.NoError(t, err)
	assert.Equal(t, string(body), result)
}

func Test_Encrypt_V12(t *testing.T) {
	password := []byte("asdf")
	body := []byte("secret")
	var encrypted string
	var err error
	encrypted, err = Encrypt(&EncryptOptions{
		Body:     &body,
		Password: &password,
		VaultID:  "test",
	})
	assert.NoError(t, err)

	var result string
	data := []byte(encrypted)
	result, err = Decrypt(&DecryptOptions{
		Data:     &data,
		Password: &password,
	})
	assert.NoError(t, err)
	assert.Equal(t, string(body), result)
}

func Test_Encrypt_V12_Bad_VaultID(t *testing.T) {
	password := []byte("asdf")
	body := []byte("secret")
	var err error
	_, err = Encrypt(&EncryptOptions{
		Body:     &body,
		Password: &password,
		VaultID:  "A;Bad;Label",
	})
	if assert.Error(t, err) {
		assert.Equal(t, errors.New("vaultID (A;Bad;Label) cannot contain ';'"), err)
	}
}

func Test_encryptV11(t *testing.T) {
	password := []byte("asdf")
	body := []byte("secret")
	var encrypted string
	var err error
	encrypted, err = encryptV11(&EncryptOptions{
		Body:     &body,
		Password: &password,
	})
	assert.NoError(t, err)

	var result string
	data := []byte(encrypted)
	result, err = Decrypt(&DecryptOptions{
		Data:     &data,
		Password: &password,
	})
	assert.NoError(t, err)
	assert.Equal(t, string(body), result)
}

func Test_encryptV12(t *testing.T) {
	password := []byte("asdf")
	body := []byte("secret")
	vaultID := "label"
	var encrypted string
	var err error
	encrypted, err = encryptV12(&EncryptOptions{
		Body:     &body,
		Password: &password,
		VaultID:  vaultID,
	})
	assert.NoError(t, err)

	var result string
	data := []byte(encrypted)
	result, err = Decrypt(&DecryptOptions{
		Data:     &data,
		Password: &password,
	})
	assert.NoError(t, err)
	assert.Equal(t, string(body), result)
}

func Test_EncryptFile_V11(t *testing.T) {
	password := []byte("asdf")
	encrypted, err := EncryptFile(&EncryptFileOptions{
		Filename: "./testdata/encrypt_file.log",
		Password: &password,
	})
	assert.NoError(t, err)
	assert.Contains(t, encrypted, "$ANSIBLE_VAULT;1.1;AES256")

	var result string
	data := []byte(encrypted)
	result, err = Decrypt(&DecryptOptions{
		Data:     &data,
		Password: &password,
	})
	assert.NoError(t, err)

	expected := `This is a test.

I have data.
`
	assert.Equal(t, expected, result)
}

func Test_EncryptFile_V12(t *testing.T) {
	password := []byte("asdf")
	encrypted, err := EncryptFile(&EncryptFileOptions{
		Filename: "./testdata/encrypt_file.log",
		Password: &password,
		VaultID:  "test-label",
	})
	assert.NoError(t, err)
	assert.Contains(t, encrypted, "$ANSIBLE_VAULT;1.2;AES256;test-label")

	var result string
	data := []byte(encrypted)
	result, err = Decrypt(&DecryptOptions{
		Data:     &data,
		Password: &password,
	})
	assert.NoError(t, err)

	expected := `This is a test.

I have data.
`
	assert.Equal(t, expected, result)
}


helpers_test.go
package avtool

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func Test_checkVaultID(t *testing.T) {
	var err error
	err = checkVaultID("")
	assert.NoError(t, err)

	err = checkVaultID("1-)90$#98klascalkkDADQXASdasd=-=+_+_=-=")
	assert.NoError(t, err)

	err = checkVaultID("a;b")
	assert.Error(t, err)
}

func Test_GenerateRandomBytes(t *testing.T) {
	var bytes []byte
	var err error
	bytes, err = GenerateRandomBytes(10)
	assert.NoError(t, err)
	assert.Equal(t, 10, len(bytes))

	bytes, err = GenerateRandomBytes(0)
	assert.NoError(t, err)
	assert.Equal(t, 0, len(bytes))
}


decrypt.go
package avtool

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"golang.org/x/crypto/pbkdf2"
	"io/ioutil"
	"strings"
)

// DecryptFileOptions is the interface used to pass data to the DecryptFile method
type DecryptFileOptions struct {
	Filename string
	Password *[]byte
}

// DecryptFile reads content of filename, decrypts it and returns string
func DecryptFile(opts *DecryptFileOptions) (result string, err error) {
	data, err := ioutil.ReadFile(opts.Filename)
	if err != nil {
		return "", err
	}
	result, err = Decrypt(&DecryptOptions{
		Data:     &data,
		Password: opts.Password,
	})
	return
}

// DecryptOptions is the interface used to pass data to the Decrypt method
type DecryptOptions struct {
	Data     *[]byte
	Password *[]byte
}

// Decrypt a string containing the ansible vault
func Decrypt(opts *DecryptOptions) (result string, err error) {
	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("ERROR: %v", r)
		}
	}()
	data := replaceCarriageReturn(string(*opts.Data))
	body := splitHeader([]byte(data))
	salt, cryptedHmac, ciphertext, err := decodeData(body)
	if err != nil {
		return "", err
	}

	key1, key2, iv := genKeyInitctr(opts.Password, salt)
	err = checkDigest(key2, cryptedHmac, ciphertext)
	if err != nil {
		return "", err
	}

	aesCipher, err := aes.NewCipher(key1)
	if err != nil {
		return "", err
	}

	aesBlock := cipher.NewCTR(aesCipher, iv)
	plaintext := make([]byte, len(ciphertext))
	aesBlock.XORKeyStream(plaintext, ciphertext)
	padding := int(plaintext[len(plaintext)-1])
	result = string(plaintext[:len(plaintext)-padding])
	return
}

// replaceCarriageReturn in order to support vault files with windows line endings
func replaceCarriageReturn(data string) string {
	return strings.ReplaceAll(data, "\r", "")
}

/*
See _split_header function
https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L288
*/
func splitHeader(data []byte) string {
	contents := string(data)
	lines := strings.Split(contents, "\n")
	header := strings.Split(lines[0], ";")
	cipherName := strings.TrimSpace(header[2])
	if cipherName != "AES256" {
		panic(fmt.Errorf("unsupported cipher: %s", cipherName))
	}
	body := strings.Join(lines[1:], "")
	return body
}

/*
See decrypt function (in class VaultAES256)
https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L741
*/
func decodeData(body string) (salt, cryptedHmac, cipherText []byte, err error) {
	decoded, err := hex.DecodeString(body)
	if err != nil {
		return nil, nil, nil, err
	}
	elements := strings.SplitN(string(decoded), "\n", 3)

	salt, err = hex.DecodeString(elements[0])
	if err != nil {
		return nil, nil, nil, err
	}

	cryptedHmac, err = hex.DecodeString(elements[1])
	if err != nil {
		return nil, nil, nil, err
	}

	cipherText, err = hex.DecodeString(elements[2])
	if err != nil {
		return nil, nil, nil, err
	}

	return
}

/*
See function _gen_key_initctr (in class VaultAES256)
https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L685
*/
func genKeyInitctr(password *[]byte, salt []byte) (key1, key2, iv []byte) {
	keyLength := 32
	ivLength := 16
	key := pbkdf2.Key(*password, salt, 10000, 2*keyLength+ivLength, sha256.New)
	key1 = key[:keyLength]
	key2 = key[keyLength:(keyLength * 2)]
	iv = key[(keyLength * 2) : (keyLength*2)+ivLength]
	return
}

/*
See decrypt function (in class VaultAES256)
https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L741
*/
func checkDigest(key2, cryptedHmac, ciphertext []byte) error {
	hmacDecrypt := hmac.New(sha256.New, key2)
	_, err := hmacDecrypt.Write(ciphertext)
	if err != nil {
		return err
	}
	expectedMAC := hmacDecrypt.Sum(nil)
	if !hmac.Equal(cryptedHmac, expectedMAC) {
		return errors.New("digests do not match - exiting")
	}
	return nil
}


encrypt.go
package avtool

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"io/ioutil"
	"strings"
)

// EncryptFileOptions is the interface used to pass data to the EncryptFile method
type EncryptFileOptions struct {
	Filename string
	Password *[]byte
	VaultID  string
}

// EncryptFile reads content of filename provided and returns encrypted string
func EncryptFile(opts *EncryptFileOptions) (result string, err error) {
	var data []byte
	data, err = ioutil.ReadFile(opts.Filename)
	if err != nil {
		return "", err
	}
	if opts.VaultID == "" {
		result, err = encryptV11(&EncryptOptions{
			Body:     &data,
			Password: opts.Password,
		})
	} else {
		result, err = encryptV12(&EncryptOptions{
			Body:     &data,
			Password: opts.Password,
			VaultID:  opts.VaultID,
		})
	}
	return
}

// EncryptOptions is the interface used to pass data to the Encrypt method
type EncryptOptions struct {
	Body     *[]byte
	Password *[]byte
	VaultID  string
}

// Encrypt will vault encrypt a piece of data.
//
// If EncryptOptions.VaultID is set, it will upversion to 1.2, otherwise it will
// default to using 1.1.
//
// EncryptOptions.VaultID must not include `;`. If it does, an error will be thrown.
func Encrypt(opts *EncryptOptions) (result string, err error) {
	err = checkVaultID(opts.VaultID)
	if err != nil {
		return "", err
	}
	if opts.VaultID == "" {
		return encryptV11(opts)
	}
	return encryptV12(opts)
}

// encryptData prepares data, encrypts string for formatting
// see https://github.com/ansible/ansible/blob/0f95371131cd41d97ad95c4e8bd983081eb29a2a/lib/ansible/parsing/vault/__init__.py#L581
func encryptData(opts *EncryptOptions) (vaultText string, err error) {
	var salt []byte
	salt, err = GenerateRandomBytes(32)
	if err != nil {
		return "", err
	}
	// salt_64 := "2262970e2309d5da757af6c473b0ed3034209cc0d48a3cc3d648c0b174c22fde"
	// salt,_ = hex.DecodeString(salt_64)
	key1, key2, iv := genKeyInitctr(opts.Password, salt)

	var cipherText []byte
	cipherText, err = createCipherText(opts.Body, key1, iv)
	if err != nil {
		return "", err
	}

	combined, err := combineParts(cipherText, key2, salt)
	if err != nil {
		return "", err
	}

	return hex.EncodeToString([]byte(combined)), nil
}

// encryptV11 encrypts data and formats the output in Ansible v1.1 vault format
// see https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L710
func encryptV11(opts *EncryptOptions) (result string, err error) {
	var encryptedData string
	encryptedData, err = encryptData(opts)
	if err != nil {
		return "", err
	}
	result = formatOutputV11(encryptedData)
	return
}

// encryptV12 encrypts data and formats the output in Ansible v1.2 vault format
// see https://docs.ansible.com/ansible/latest/user_guide/vault.html#ansible-vault-payload-format-1-1-1-2
// see https://github.com/ansible/ansible/blob/0f95371131cd41d97ad95c4e8bd983081eb29a2a/lib/ansible/parsing/vault/__init__.py#L581
func encryptV12(opts *EncryptOptions) (result string, err error) {
	var encryptedData string
	encryptedData, err = encryptData(opts)
	if err != nil {
		return "", err
	}
	result = formatOutputV12(encryptedData, opts.VaultID)
	return
}

func createCipherText(body *[]byte, key1, iv []byte) ([]byte, error) {
	bs := aes.BlockSize
	padding := bs - len(string(*body))%bs
	if padding == 0 {
		padding = bs
	}
	padChar := rune(padding)
	padArray := make([]byte, padding)
	for i := range padArray {
		padArray[i] = byte(padChar)
	}
	plaintext := *body
	plaintext = append(plaintext, padArray...)

	aesCipher, err := aes.NewCipher(key1)
	if err != nil {
		return nil, err
	}
	ciphertext := make([]byte, len(plaintext))

	aesBlock := cipher.NewCTR(aesCipher, iv)
	aesBlock.XORKeyStream(ciphertext, plaintext)
	return ciphertext, nil
}

func combineParts(ciphertext, key2, salt []byte) (string, error) {
	hmacEncrypt := hmac.New(sha256.New, key2)
	_, err := hmacEncrypt.Write(ciphertext)
	if err != nil {
		return "", err
	}
	hexSalt := hex.EncodeToString(salt)
	hexHmac := hmacEncrypt.Sum(nil)
	hexCipher := hex.EncodeToString(ciphertext)
	// nolint:unconvert
	combined := string(hexSalt) + "\n" + hex.EncodeToString([]byte(hexHmac)) + "\n" + string(hexCipher)
	return combined, nil
}

// https://github.com/ansible/ansible/blob/0b8011436dc7f842b78298848e298f2a57ee8d78/lib/ansible/parsing/vault/__init__.py#L268
func formatOutputV11(vaultText string) string {
	heading := "$ANSIBLE_VAULT"
	version := "1.1"
	cipherName := "AES256"

	headerElements := make([]string, 3)
	headerElements[0] = heading
	headerElements[1] = version
	headerElements[2] = cipherName
	header := strings.Join(headerElements, ";")

	elements := make([]string, 1)
	elements[0] = header
	for i := 0; i < len(vaultText); i += 80 {
		end := i + 80
		if end > len(vaultText) {
			end = len(vaultText)
		}
		elements = append(elements, vaultText[i:end])
	}
	elements = append(elements, "")

	whole := strings.Join(elements, "\n")
	return whole
}

func formatOutputV12(vaultText, vaultID string) string {
	heading := "$ANSIBLE_VAULT"
	version := "1.2"
	cipherName := "AES256"

	headerElements := make([]string, 4)
	headerElements[0] = heading
	headerElements[1] = version
	headerElements[2] = cipherName
	headerElements[3] = vaultID
	header := strings.Join(headerElements, ";")

	elements := make([]string, 1)
	elements[0] = header
	for i := 0; i < len(vaultText); i += 80 {
		end := i + 80
		if end > len(vaultText) {
			end = len(vaultText)
		}
		elements = append(elements, vaultText[i:end])
	}
	elements = append(elements, "")

	whole := strings.Join(elements, "\n")
	return whole
}
